statsmodels/gam/tests/test_gam.py

```diff
@@ -0,0 +1,791 @@
+    # plt.plot(y0, label='y0')
+    # plt.plot(y, '.', label='y')
+    # plt.legend(loc='best')
+    # plt.show()
+
+    assert_allclose(y_est, y0, atol=0.04)
+
+
+def test_glm_pirls_compatibility():
+    np.random.seed(0)
+
+    n = 500
+    x1 = np.linspace(-3, 3, n)
+    x2 = np.random.rand(n)
+
+    x = np.vstack([x1, x2]).T
+    y1 = np.sin(x1) / x1
+    y2 = x2 * x2
+    y0 = y1 + y2
+    y = y0 + np.random.normal(0, .3, n)
+    y -= y.mean()
+    y0 -= y0.mean()
+
+    # TODO: we have now alphas == alphas_glm
+    alphas = [5.75] * 2
+    alphas_glm = [1.2] * 2
+    # using constraints avoids singular exog.
+    cs = BSplines(x, df=[10, 10], degree=[3, 3], constraints='center')
+
+    gam_pirls = GLMGam(y, smoother=cs, alpha=alphas)
+    gam_glm = GLMGam(y, smoother=cs, alpha=alphas)
+
+    gam_res_glm = gam_glm.fit(method='nm', max_start_irls=0,
+                              disp=1, maxiter=20000, maxfun=10000)
+    gam_res_glm = gam_glm.fit(start_params=gam_res_glm.params,
+                              method='bfgs', max_start_irls=0,
+                              disp=1, maxiter=20000, maxfun=10000)
+    gam_res_pirls = gam_pirls.fit()
+
+    y_est_glm = np.dot(cs.basis, gam_res_glm.params)
+    y_est_glm -= y_est_glm.mean()
+    y_est_pirls = np.dot(cs.basis, gam_res_pirls.params)
+    y_est_pirls -= y_est_pirls.mean()
+
+    # plt.plot(y_est_pirls)
+    # plt.plot(y_est_glm)
+    # plt.plot(y, '.')
+    # plt.show()
+    assert_allclose(gam_res_glm.params, gam_res_pirls.params, atol=5e-5,
+                    rtol=5e-5)
+    assert_allclose(y_est_glm, y_est_pirls, atol=5e-5)
+
+
+def test_zero_penalty():
+    x, y, poly = multivariate_sample_data()
+    alphas = [0, 0]
+    gam_gs = GLMGam(y, smoother=poly, alpha=alphas)
+    gam_gs_res = gam_gs.fit()
+    y_est_gam = gam_gs_res.predict()
+
+    glm = GLM(y, poly.basis).fit()
+    y_est = glm.predict()
+
+    assert_allclose(y_est, y_est_gam)
+
+
+def test_spl_s():
+    # matrix from R
+    spl_s_R = [[0, 0, 0.000000000, 0.000000000, 0.000000000, 0.000000000],
+               [0, 0, 0.000000000, 0.000000000, 0.000000000, 0.000000000],
+               [0, 0, 0.001400000, 0.000200000, -0.001133333, -0.001000000],
+               [0, 0, 0.000200000, 0.002733333, 0.001666667, -0.001133333],
+               [0, 0, -0.001133333, 0.001666667, 0.002733333, 0.000200000],
+               [0, 0, -0.001000000, -0.001133333, 0.000200000, 0.001400000]]
+
+    np.random.seed(1)
+    x = np.random.normal(0, 1, 10)
+    xk = np.array([0.2, .4, .6, .8])
+    cs = UnivariateCubicSplines(x, df=4)
+    cs.knots = xk
+
+    spl_s = cs._splines_s()
+    assert_allclose(spl_s_R, spl_s, atol=4.e-10)
+
+
+def test_partial_values2():
+    np.random.seed(0)
+    n = 1000
+    x = np.random.uniform(0, 1, (n, 2))
+    x = x - x.mean()
+    y = x[:, 0] * x[:, 0] + np.random.normal(0, .01, n)
+    y -= y.mean()
+    alpha = 0.0
+    # BUG: mask is incorrect if exog is not None, start_idx missing
```

Comment:
is "BUG" used frequently?  I commonly grep for "TODO" or "FIXME" when looking for things that need attention.

Reply:
this needs another look and verifying what the current code does.
Right now I'm just looking at easy comments/changes.